File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/chunk/ClientChunk.java
package pize.tests.voxelgame.client.chunk;

import pize.math.vecmath.matrix.Matrix4f;
import pize.tests.voxelgame.client.chunk.mesh.ChunkMeshStack;
import pize.tests.voxelgame.client.control.camera.GameCamera;
import pize.tests.voxelgame.main.chunk.LevelChunk;
import pize.tests.voxelgame.main.chunk.storage.ChunkPos;
import pize.tests.voxelgame.main.chunk.storage.HeightmapType;
import pize.tests.voxelgame.main.level.ChunkManagerUtils;
import pize.tests.voxelgame.main.block.BlockData;
import pize.tests.voxelgame.client.block.Blocks;
import pize.tests.voxelgame.client.level.ClientLevel;

import static pize.tests.voxelgame.main.chunk.ChunkUtils.isOutOfBounds;

public class ClientChunk extends LevelChunk{

    private final ChunkMeshStack meshStack;
    private final Matrix4f translationMatrix;
    
    public ClientChunk(ClientLevel level, ChunkPos position){
        super(level, position);
        meshStack = new ChunkMeshStack();
        translationMatrix = new Matrix4f();
    }


    public ChunkMeshStack getMeshStack(){
        return meshStack;
    }


    public Matrix4f getTranslationMatrix(){
        return translationMatrix;
    }

    public void updateTranslationMatrix(GameCamera camera){
        translationMatrix.toTranslated(
            position.globalX() - camera.getX(),
            0,
            position.globalZ() - camera.getZ()
        );
    }
    
    
    public boolean setBlock(int lx, int y, int lz, short blockState){
        if(!super.setBlock(lx, y, lz, blockState) || isOutOfBounds(lx, lz))
            return false;
        
        getHeightMap(HeightmapType.HIGHEST).update(lx, y, lz, BlockData.getID(blockState) != Blocks.AIR.getID());
        rebuild(true);
        ChunkManagerUtils.rebuildNeighborChunks(this, lx, lz);
        
        return true;
    }
    
    public void setLight(int lx, int y, int lz, int level){
        super.setLight(lx, y, lz, level);
        rebuild(true);
    }
    
    
    public void rebuild(boolean important){
        getLevel().getChunkManager().rebuildChunk(this, important);
    }
    
    public ClientLevel getLevel(){
        return (ClientLevel) level;
    }
    
    public ClientChunk getNeighbor(int chunkX, int chunkZ){
        return getLevel().getChunkManager().getChunk(position.x + chunkX, position.z + chunkZ);
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/chunk/ClientChunk.kt
package pize.tests.voxelgame.client.chunk

import pize.math.vecmath.matrix.Matrix4f
import pize.tests.voxelgame.client.block.Blocks
import pize.tests.voxelgame.client.chunk.mesh.ChunkMeshStack
import pize.tests.voxelgame.client.control.camera.GameCamera
import pize.tests.voxelgame.client.level.ClientLevel
import pize.tests.voxelgame.main.block.BlockData
import pize.tests.voxelgame.main.chunk.ChunkUtils
import pize.tests.voxelgame.main.chunk.LevelChunk
import pize.tests.voxelgame.main.chunk.storage.ChunkPos
import pize.tests.voxelgame.main.chunk.storage.HeightmapType
import pize.tests.voxelgame.main.level.ChunkManagerUtils
import pize.tests.voxelgame.main.level.Level

class ClientChunk(level: ClientLevel, position: ChunkPos?) : LevelChunk(level, position) {
    val meshStack: ChunkMeshStack
    val translationMatrix: Matrix4f

    init {
        meshStack = ChunkMeshStack()
        translationMatrix = Matrix4f()
    }

    fun updateTranslationMatrix(camera: GameCamera) {
        translationMatrix.toTranslated(
            position!!.globalX() - camera.x,
            0f,
            position.globalZ() - camera.z
        )
    }

    override fun setBlock(lx: Int, y: Int, lz: Int, blockState: Short): Boolean {
        if (!super.setBlock(lx, y, lz, blockState) || ChunkUtils.isOutOfBounds(lx, lz)) return false
        getHeightMap(HeightmapType.HIGHEST)!!.update(lx, y, lz, BlockData.getID(blockState) != Blocks.AIR.id)
        rebuild(true)
        ChunkManagerUtils.rebuildNeighborChunks(this, lx, lz)
        return true
    }

    override fun setLight(lx: Int, y: Int, lz: Int, level: Int) {
        super.setLight(lx, y, lz, level)
        rebuild(true)
    }

    fun rebuild(important: Boolean) {
        this.level.getChunkManager().rebuildChunk(this, important)
    }

    override val level: Level
        get() = level as ClientLevel

    override fun getNeighbor(chunkX: Int, chunkZ: Int): ClientChunk? {
        return this.level.getChunkManager().getChunk(position!!.x + chunkX, position.z + chunkZ)
    }
}
